unblob/handlers/filesystem/yaffs/utils.py

```diff
@@ -0,0 +1,572 @@
+            )
+        except NodeIDAbsentError:
+            pass
+        except DuplicatedNodeIdError:
+            pass
+
+    def get_entry(self, object_id: int) -> Optional[YAFFSEntry]:
+        try:
+            entry = self.file_entries.get_node(object_id)
+            if entry:
+                return entry.data
+            else:
+                return None
+        except NodeIDAbsentError:
+            pass
+
+    def resolve_path(self, entry: YAFFSEntry) -> Path:
+        resolved_path = Path(entry.name)
+        parent = self.file_entries[entry.parent_obj_id].data
+        if parent is not None:
+            return self.resolve_path(parent).joinpath(resolved_path)
+        else:
+            return resolved_path
+
+    def get_file_bytes(self, entry: YAFFSEntry) -> Iterable[bytes]:
+        for chunk in entry.chunks:
+            self.file.seek(
+                entry.start_offset
+                + ((chunk.id - 1) * (self.config.page_size + self.config.spare_size)),
+                io.SEEK_SET,
+            )
+            yield self.file.read(chunk.byte_count)
+
+    def extract(self, outdir: Path):
+        self.file_entries.show()
+        for entry in [
+            self.file_entries.get_node(node).data
+            for node in self.file_entries.expand_tree(mode=Tree.DEPTH)
+        ]:
+            if entry is None:
+                continue
+            self.extract_entry(entry, outdir)
+
+    def extract_entry(self, entry: YAFFSEntry, outdir: Path):  # noqa: C901
+        entry_path = self.resolve_path(entry)
+
+        if not is_safe_path(outdir, entry_path):
+            logger.warning(
+                "Potential path traversal attempt", outdir=outdir, path=entry_path
+            )
+            return
+
+        out_path = outdir.joinpath(entry_path)
+
+        if entry.type == YAFFS_OBJECT_TYPE.DIRECTORY:
+            logger.debug("creating directory", dir_path=out_path, _verbosity=3)
+            out_path.mkdir(exist_ok=True)
+        elif entry.type == YAFFS_OBJECT_TYPE.FILE:
+            logger.debug("creating file", file_path=out_path, _verbosity=3)
+            with out_path.open("wb") as f:
+                for chunk in self.get_file_bytes(entry):
+                    f.write(chunk)
+        elif entry.type == YAFFS_OBJECT_TYPE.SPECIAL and (os.geteuid() == 0):
+            logger.debug("creating special file", special_path=out_path, _verbosity=3)
+            os.mknod(out_path.as_posix(), entry.yst_mode, entry.yst_rdev)
+        elif entry.type == YAFFS_OBJECT_TYPE.SYMLINK:
+            logger.debug("creating symlink", file_path=out_path, _verbosity=3)
+            out_path.symlink_to(Path(entry.alias))
+        elif entry.type == YAFFS_OBJECT_TYPE.HARDLINK:
+            logger.debug("creating hardlink", file_path=out_path, _verbosity=3)
+            src_entry = self.file_entries[entry.equiv_id]
+            src_path = self.resolve_path(src_entry)
+            src_path.link_to(out_path)
```

Comment:
can here be a traversal attempt?

Reply:
Indeed, in a similar way to https://bugzilla.redhat.com/show_bug.cgi?id=CVE-2021-41072

Reply:
Fixed